File: quantum-web/quantum-web-api/src/main/java/com/quantum/web/exception/GlobalExceptionHandler.java
package com.quantum.web.exception;

import com.quantum.web.controller.KiwoomAccountController;
import com.quantum.web.service.KiwoomAccountManagementService;
import com.quantum.web.service.KiwoomTokenService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.validation.BindingResult;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.ConstraintViolationException;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * 글로벌 예외 처리 핸들러
 * 
 * 모든 REST API 예외를 일관된 형식으로 처리하고 응답
 * - 키움증권 계정 관리 예외
 * - 인증/인가 예외
 * - 유효성 검증 예외
 * - 시스템 예외
 */
@RestControllerAdvice
@Slf4j
public class GlobalExceptionHandler {

    /**
     * 키움증권 계정 관리 예외 처리
     */
    @ExceptionHandler(KiwoomAccountManagementService.KiwoomAccountManagementException.class)
    public ResponseEntity<ErrorResponse> handleKiwoomAccountManagementException(
            KiwoomAccountManagementService.KiwoomAccountManagementException ex,
            WebRequest request) {
        
        log.warn("Kiwoom account management error: {}", ex.getMessage(), ex);
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message(ex.getMessage())
                .errorCode("KIWOOM_ACCOUNT_ERROR")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse);
    }

    /**
     * 키움증권 토큰 관리 예외 처리
     */
    @ExceptionHandler(KiwoomTokenService.TokenManagementException.class)
    public ResponseEntity<ErrorResponse> handleTokenManagementException(
            KiwoomTokenService.TokenManagementException ex,
            WebRequest request) {
        
        log.warn("Kiwoom token management error: {}", ex.getMessage(), ex);
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message(ex.getMessage())
                .errorCode("TOKEN_MANAGEMENT_ERROR")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse);
    }

    /**
     * 인증 예외 처리
     */
    @ExceptionHandler(AuthenticationException.class)
    public ResponseEntity<ErrorResponse> handleAuthenticationException(
            AuthenticationException ex,
            WebRequest request) {
        
        log.warn("Authentication error: {}", ex.getMessage());
        
        String message = ex instanceof BadCredentialsException 
            ? "Invalid username or password" 
            : "Authentication failed";
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message(message)
                .errorCode("AUTHENTICATION_ERROR")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(errorResponse);
    }

    /**
     * 접근 권한 예외 처리
     */
    @ExceptionHandler(AccessDeniedException.class)
    public ResponseEntity<ErrorResponse> handleAccessDeniedException(
            AccessDeniedException ex,
            WebRequest request) {
        
        log.warn("Access denied: {}", ex.getMessage());
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message("Access denied. Insufficient privileges.")
                .errorCode("ACCESS_DENIED")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(errorResponse);
    }

    /**
     * 요청 본문 유효성 검증 실패 처리
     */
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(
            MethodArgumentNotValidException ex,
            WebRequest request) {
        
        log.debug("Validation error: {}", ex.getMessage());
        
        BindingResult bindingResult = ex.getBindingResult();
        Map<String, String> fieldErrors = new HashMap<>();
        
        for (FieldError fieldError : bindingResult.getFieldErrors()) {
            fieldErrors.put(fieldError.getField(), fieldError.getDefaultMessage());
        }
        
        String message = "Validation failed for " + fieldErrors.size() + " field(s)";
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message(message)
                .errorCode("VALIDATION_ERROR")
                .data(fieldErrors)
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse);
    }

    /**
     * 경로 매개변수 유효성 검증 실패 처리
     */
    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<ErrorResponse> handleConstraintViolationException(
            ConstraintViolationException ex,
            WebRequest request) {
        
        log.debug("Constraint violation: {}", ex.getMessage());
        
        String violations = ex.getConstraintViolations()
                .stream()
                .map(ConstraintViolation::getMessage)
                .collect(Collectors.joining(", "));
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message("Validation failed: " + violations)
                .errorCode("CONSTRAINT_VIOLATION")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse);
    }

    /**
     * 잘못된 인수 예외 처리
     */
    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ErrorResponse> handleIllegalArgumentException(
            IllegalArgumentException ex,
            WebRequest request) {
        
        log.debug("Illegal argument: {}", ex.getMessage());
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message("Invalid request: " + ex.getMessage())
                .errorCode("INVALID_ARGUMENT")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse);
    }

    /**
     * 일반적인 런타임 예외 처리
     */
    @ExceptionHandler(RuntimeException.class)
    public ResponseEntity<ErrorResponse> handleRuntimeException(
            RuntimeException ex,
            WebRequest request) {
        
        log.error("Runtime error: {}", ex.getMessage(), ex);
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message("An unexpected error occurred. Please try again later.")
                .errorCode("RUNTIME_ERROR")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
    }

    /**
     * 모든 예외의 최종 처리
     */
    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleGenericException(
            Exception ex,
            WebRequest request) {
        
        log.error("Unexpected error: {}", ex.getMessage(), ex);
        
        ErrorResponse errorResponse = ErrorResponse.builder()
                .success(false)
                .message("An internal server error occurred. Please contact support.")
                .errorCode("INTERNAL_SERVER_ERROR")
                .timestamp(Instant.now())
                .path(getPath(request))
                .build();
        
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
    }

    /**
     * 요청 경로 추출
     */
    private String getPath(WebRequest request) {
        String path = request.getDescription(false);
        return path.startsWith("uri=") ? path.substring(4) : path;
    }

    /**
     * 에러 응답 DTO
     */
    @lombok.Data
    @lombok.Builder
    public static class ErrorResponse {
        private boolean success;
        private String message;
        private String errorCode;
        private Object data;
        private Instant timestamp;
        private String path;
    }
}

File: quantum-web-api/src/main/kotlin/com/quantum/config/GlobalExceptionHandler.kt
package com.quantum.config

import com.quantum.common.dto.ErrorCodes
import com.quantum.common.dto.StandardErrorResponse
import com.quantum.common.dto.UserMessages
import com.quantum.stock.application.service.StockDataException
import com.quantum.user.presentation.dto.ErrorResponse
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {
    
    private val logger = LoggerFactory.getLogger(this::class.java)
    
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationExceptions(ex: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
        val errors = ex.bindingResult.allErrors.joinToString(", ") { error ->
            when (error) {
                is FieldError -> "${error.field}: ${error.defaultMessage}"
                else -> error.defaultMessage ?: "Unknown validation error"
            }
        }

        logger.warn("Validation error: $errors")
        
        return ResponseEntity.badRequest().body(
            ErrorResponse(
                error = "VALIDATION_ERROR",
                message = errors
            )
        )
    }
    
    @ExceptionHandler(BadCredentialsException::class)
    fun handleBadCredentialsException(ex: BadCredentialsException): ResponseEntity<ErrorResponse> {
        logger.warn("Authentication failed: ${ex.message}")
        
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(
            ErrorResponse(
                error = "BAD_CREDENTIALS",
                message = "이메일 또는 비밀번호가 올바르지 않습니다."
            )
        )
    }
    
    @ExceptionHandler(UsernameNotFoundException::class)
    fun handleUsernameNotFoundException(ex: UsernameNotFoundException): ResponseEntity<ErrorResponse> {
        logger.warn("User not found: ${ex.message}")
        
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(
            ErrorResponse(
                error = "USER_NOT_FOUND",
                message = "사용자를 찾을 수 없습니다."
            )
        )
    }
    
    @ExceptionHandler(IllegalArgumentException::class)
    fun handleIllegalArgumentException(ex: IllegalArgumentException): ResponseEntity<ErrorResponse> {
        logger.warn("Illegal argument: ${ex.message}")
        
        return ResponseEntity.badRequest().body(
            ErrorResponse(
                error = "ILLEGAL_ARGUMENT",
                message = ex.message ?: "잘못된 요청입니다."
            )
        )
    }
    
    @ExceptionHandler(IllegalStateException::class)
    fun handleIllegalStateException(ex: IllegalStateException): ResponseEntity<ErrorResponse> {
        logger.warn("Illegal state: ${ex.message}")
        
        return ResponseEntity.status(HttpStatus.CONFLICT).body(
            ErrorResponse(
                error = "ILLEGAL_STATE",
                message = ex.message ?: "요청을 처리할 수 없는 상태입니다."
            )
        )
    }
    
    @ExceptionHandler(MissingServletRequestParameterException::class)
    fun handleMissingRequestParameter(ex: MissingServletRequestParameterException): ResponseEntity<ErrorResponse> {
        logger.warn("Missing request parameter: ${ex.parameterName}")
        
        val message = when (ex.parameterName) {
            "environment" -> "필수 파라미터 'environment'가 누락되었습니다. LIVE 또는 SANDBOX 중 선택해주세요."
            else -> "필수 파라미터 '${ex.parameterName}'가 누락되었습니다."
        }
        
        return ResponseEntity.badRequest().body(
            ErrorResponse(
                error = "MISSING_PARAMETER",
                message = message
            )
        )
    }
    
    @ExceptionHandler(MethodArgumentTypeMismatchException::class)
    fun handleTypeMismatch(ex: MethodArgumentTypeMismatchException): ResponseEntity<ErrorResponse> {
        logger.warn("Type mismatch for parameter: ${ex.name}")
        
        val message = when (ex.name) {
            "environment" -> "environment 파라미터는 LIVE 또는 SANDBOX 값만 허용됩니다."
            else -> "파라미터 '${ex.name}'의 값이 올바르지 않습니다."
        }
        
        return ResponseEntity.badRequest().body(
            ErrorResponse(
                error = "INVALID_PARAMETER",
                message = message
            )
        )
    }
    
    // 주식 데이터 서비스 예외 처리 (표준화된 응답)
    @ExceptionHandler(StockDataException::class)
    fun handleStockDataException(
        ex: StockDataException, 
        request: HttpServletRequest
    ): ResponseEntity<StandardErrorResponse> {
        logger.error("Stock data service error: ${ex.message}", ex)
        
        val errorResponse = StandardErrorResponse.dataNotFound(
            message = ex.message ?: UserMessages.STOCK_NOT_FOUND,
            path = request.requestURI
        )
        
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse)
    }
    
    // KIS API 관련 오류 처리 (표준화된 응답)
    @ExceptionHandler(RuntimeException::class)
    fun handleKisApiException(
        ex: RuntimeException,
        request: HttpServletRequest
    ): ResponseEntity<StandardErrorResponse> {
        val message = ex.message ?: ""
        
        // KIS API 관련 오류 패턴 감지
        when {
            message.contains("KIS", ignoreCase = true) || 
            message.contains("token", ignoreCase = true) ||
            message.contains("API", ignoreCase = true) -> {
                logger.error("KIS API error occurred: ${ex.message}", ex)
                
                val errorResponse = StandardErrorResponse.kisApiError(
                    UserMessages.KIS_API_CONNECTION_FAILED
                ).copy(path = request.requestURI)
                
                return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE).body(errorResponse)
            }
            message.contains("데이터", ignoreCase = true) ||
            message.contains("data", ignoreCase = true) -> {
                logger.error("Data processing error: ${ex.message}", ex)
                
                val errorResponse = StandardErrorResponse(
                    error = ErrorCodes.DATA_INTEGRITY_ERROR,
                    message = UserMessages.CHART_DATA_NOT_AVAILABLE,
                    details = "데이터 처리 중 오류가 발생했습니다.",
                    path = request.requestURI
                )
                
                return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(errorResponse)
            }
            else -> {
                logger.error("Runtime error occurred: ${ex.message}", ex)
                
                val errorResponse = StandardErrorResponse.internalServerError(
                    UserMessages.SERVER_ERROR
                ).copy(path = request.requestURI)
                
                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse)
            }
        }
    }
    
    // 일반적인 오류는 더 구체적인 정보 없이 처리 (표준화된 응답)
    @ExceptionHandler(Exception::class)
    fun handleGenericException(
        ex: Exception,
        request: HttpServletRequest
    ): ResponseEntity<StandardErrorResponse> {
        logger.error("Unexpected error occurred", ex)
        
        // 절대 더미 데이터를 생성하거나 성공으로 위장하지 않음
        val errorResponse = StandardErrorResponse.internalServerError(
            UserMessages.SERVER_ERROR
        ).copy(
            details = "예기치 않은 오류가 발생했습니다.",
            path = request.requestURI
        )
        
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse)
    }
}